### PR TITLE
[rllib] [experimental] Decentralized Distributed PPO for torch (DD-PPO)

### PR DESCRIPTION
**Pull request by ericl**
## Why are these changes needed?

This adds an experimental implementation of distributed SGD for PPO (e.g., https://arxiv.org/abs/1911.00357) using `torch.distributed.allreduce`. It learns on toy examples about as fast as non-distributed sgd.

You can try this out with `./train.py --run=PPO --env=CartPole-v0 --torch --config='{"distributed_data_parallel_optimizer": true, "num_workers": 2, "sample_batch_size": 4000, "train_batch_size": 4000, "batch_mode": "truncate_episodes"}'`

To run on GPUs you will also need to set `num_gpus_per_worker=1`.

The implementation is fairly simple:
- The data parallel policy optimizer creates a process group that includes all the rollout workers.
- Each rollout worker samples a batch of data and then performs minibatch SGD.
- We synchronize the gradients of the minibatch SGD using `torch.distributed.allreduce`.

We could also implement straggler mitigation by dynamically stopping rollouts based on global completion, which can be done as an optimization later. Right now we assume each worker produces an identical number of samples per iteration.

## Related issue number

https://github.com/ray-project/ray/issues/6636

### CODE REVIEW

File: python/ray/tune/progress_reporter.py [hunk 1]
```diff
@@ -149,7 +149,10 @@ def trial_progress_str(trials, metrics=None, fmt="psql", max_rows=20):
 
     # Pre-process trials to figure out what columns to show.
     keys = list(metrics or DEFAULT_PROGRESS_KEYS)
-    keys = [k for k in keys if any(t.last_result.get(k) for t in trials)]
+    keys = [
+        k for k in keys if any(
+            t.last_result.get(k) is not None for t in trials)
```
Comment (ericl) [inline python/ray/tune/progress_reporter.py]: Bug in tune incidentally fixed here that ignored metrics with value 0.

File: rllib/agents/ppo/ppo.py [hunk 2]
```diff
@@ -72,6 +75,33 @@
 
 
 def choose_policy_optimizer(workers, config):
+    if config["distributed_data_parallel_optimizer"]:
```
Comment (ericl) [inline rllib/agents/ppo/ppo.py]: This config is a bit awkward... we can probably revisit this once we add documentation for this.

Comment (ericl) [reply rllib/agents/ppo/ppo.py]: (i.e., we could add a DD_PPO algorithm that configures these things correctly by default).

File: rllib/optimizers/multi_gpu_optimizer.py [hunk 3]
```diff
@@ -220,11 +221,3 @@ def stats(self):
                                         3),
                 "learner": self.learner_stats,
             })
-
```
Comment (sven1977) [inline rllib/optimizers/multi_gpu_optimizer.py]: 👍

File: rllib/utils/sgd.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+        samples = MultiAgentBatch({DEFAULT_POLICY_ID: samples}, samples.count)
+
+    fetches = {}
+    for policy_id, policy in policies.items():
+        if policy_id not in samples.policy_batches:
+            continue
+
+        batch = samples.policy_batches[policy_id]
+        for field in standardize_fields:
+            value = batch[field]
```
Comment (sven1977) [inline rllib/utils/sgd.py]: Can we make a util function out of this?

File: rllib/utils/sgd.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+    while i < samples.count:
+        slices.append((i, i + sgd_minibatch_size))
+        i += sgd_minibatch_size
+    random.shuffle(slices)
+
+    for i, j in slices:
+        yield samples.slice(i, j)
+
+
+def do_minibatch_sgd(samples, policies, local_worker, num_sgd_iter,
```
Comment (sven1977) [inline rllib/utils/sgd.py]: Could you add a more detailed doctoring?
Args? Like `samples` could be a single SampleBatch, etc..

**Review by sven1977 — APPROVED** [review]

File: rllib/agents/ppo/ppo.py [hunk 6]
```diff
@@ -3,7 +3,8 @@
 from ray.rllib.agents import with_common_config
 from ray.rllib.agents.ppo.ppo_tf_policy import PPOTFPolicy
 from ray.rllib.agents.trainer_template import build_trainer
-from ray.rllib.optimizers import SyncSamplesOptimizer, LocalMultiGPUOptimizer
+from ray.rllib.optimizers import SyncSamplesOptimizer, \
+    LocalMultiGPUOptimizer, TorchDistributedDataParallelOptimizer
```
[the remark this answers is not in the document]
Comment (ericl) [reply rllib/agents/ppo/ppo.py]: I agree, though I am thinking to wait until we actually make it support both before renaming to be generic (also, I might not have time to implement/test the TF distributed version any time soon).

File: rllib/utils/sgd.py (hunk 5 above)
Comment (ericl) [reply rllib/utils/sgd.py]: Done.

File: rllib/utils/sgd.py (hunk 4 above)
Comment (ericl) [reply rllib/utils/sgd.py]: Done

File: rllib/agents/ppo/ppo.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (ericl) [reply rllib/agents/ppo/ppo.py]: Also am thinking about how policy optimizers fit into the upcoming distributed DSL, whether the DSL is just a way to define optimizers / trainers or replaces optimizers or both optimizers and trainers. Probably just the former for now.

**Review by richardliaw — APPROVED** [review]

**Pull request merged**